packages/abstract-core/src/objects/gov_type.rs

```diff
@@ -1,54 +1,141 @@
 //! # Governance structure object
 
+use cosmwasm_std::{Addr, Api};
+use cw_address_like::AddressLike;
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 
+use crate::AbstractError;
+
+const MIN_GOV_TYPE_LENGTH: usize = 4;
+const MAX_GOV_TYPE_LENGTH: usize = 64;
+
 /// Governance types
 #[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
-
-pub enum GovernanceDetails {
+#[non_exhaustive]
+pub enum GovernanceDetails<T: AddressLike> {
     /// A single address is admin
     Monarchy {
         /// The monarch's address
-        monarch: String,
+        monarch: T,
     },
     /// An external governance source
     External {
         /// The external contract address
-        governance_address: String,
+        governance_address: T,
         /// Governance type used for doing extra off-chain queries depending on the type.
         governance_type: String,
     },
 }
 
-impl ToString for GovernanceDetails {
-    fn to_string(&self) -> String {
+impl GovernanceDetails<String> {
+    /// Verify the governance details and convert to `Self<Addr>`
+    pub fn verify(self, api: &dyn Api) -> Result<GovernanceDetails<Addr>, AbstractError> {
+        match self {
+            GovernanceDetails::Monarchy { monarch } => {
+                let addr = api.addr_validate(&monarch)?;
+                Ok(GovernanceDetails::Monarchy { monarch: addr })
+            }
+            GovernanceDetails::External {
+                governance_address,
+                governance_type,
+            } => {
+                let addr = api.addr_validate(&governance_address)?;
+                if governance_type.len() < MIN_GOV_TYPE_LENGTH
+                    || governance_type.len() > MAX_GOV_TYPE_LENGTH
+                {
+                    return Err(AbstractError::FormattingError {
+                        object: "governance_type".to_string(),
+                        expected: format!(
+                            "a string of length between {} and {}",
+                            MIN_GOV_TYPE_LENGTH, MAX_GOV_TYPE_LENGTH
+                        ),
+                        actual: format!("a string of length {}", governance_type.len()),
+                    });
```

Comment:
As we are doing similar validations for length we could pull this out into a `ValidatedString` or other. 

And additionally - should we disallow spaces / other chars?

Reply:
in a name, hmm maybe allow it? Pulling it into a type is kinda overkill imo.

Reply:
true true. Point still stands about the spaces.

Reply:
we can do all kinds of verification like mentioned in the report. The question is: do we really want to include that logic in the on-chain contract? These checks are easier and cheaper to perform in the front-end and the strings (be it name, governance type, etc. don't effect anything on-chain. 

The line I would draw is that the checks should be performed on contracts that are low-use and are rarely loaded, that way the contract's binary size doesn't impact gas costs of module calls.

Reply:
format gov-type as module namespaces